vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs

```diff
@@ -0,0 +1,476 @@
+                    let! acc' = walkModuleOrNamespace acc (decls, moduleRange)
+                    return! loop (acc' @ acc) rest
+            }
+            return! loop [] modules
+        | _ -> return []
+    }
+
+    let getEffectiveOpenDeclarationsAtLocation (pos: pos) (ast: ParsedInput) =
+        let openStatements =
+            match ast with
+            | ParsedInput.ImplFile (ParsedImplFileInput(_, _, _, _, _, modules, _)) ->
+                let rec walkModuleOrNamespace openStatements (decls, moduleRange) =
+                    decls
+                    |> List.fold (fun acc -> 
+                        function
+                        | SynModuleDecl.NestedModule (_, _, nestedModuleDecls, _, nestedModuleRange) -> 
+                            if rangeContainsPos moduleRange pos then
+                                walkModuleOrNamespace acc (nestedModuleDecls, nestedModuleRange)
+                            else acc
+                        | SynModuleDecl.Open (LongIdentWithDots(longIdent, _), openDeclRange) ->
+                            let identArray = processOpenDeclaration longIdent
+                            if openDeclRange.EndLine < pos.Line || (openDeclRange.EndLine = pos.Line && openDeclRange.EndColumn < pos.Column) then 
+                                String.Join(".", identArray) :: acc
+                            else acc
+                        | _ -> acc) openStatements 
+
+                modules
+                |> List.fold (fun acc (SynModuleOrNamespace(_, _, _, decls, _, _, _, moduleRange)) ->
+                       if rangeContainsPos moduleRange pos then
+                           walkModuleOrNamespace acc (decls, moduleRange) @ acc
+                       else acc) []
+            | ParsedInput.SigFile(ParsedSigFileInput(_, _, _, _, modules)) -> 
+                let rec walkModuleOrNamespaceSig openStatements (decls, moduleRange) =
+                    decls
+                    |> List.fold (fun acc -> 
+                        function
+                        | SynModuleSigDecl.NestedModule (_, _, nestedModuleDecls, nestedModuleRange) -> 
+                            if rangeContainsPos moduleRange pos then
+                                walkModuleOrNamespaceSig acc (nestedModuleDecls, nestedModuleRange)
+                            else acc
+                        | SynModuleSigDecl.Open (longIdent, openDeclRange) ->
+                            let identArray = processOpenDeclaration longIdent
+                            if openDeclRange.EndLine < pos.Line || (openDeclRange.EndLine = pos.Line && openDeclRange.EndColumn < pos.Column) then 
+                                String.Join(".", identArray) :: acc
+                            else acc
+                        | _ -> acc) openStatements
+
+                modules
+                |> List.fold (fun acc (SynModuleOrNamespaceSig(_, _, _, decls, _, _, _, moduleRange)) ->
```

Comment:
Please remove nearly all the uses of ``fold`` in this module in favor of other combinators (as mentioned above).  They all look like filters, choose, etc.